feed/models.py

```diff
@@ -0,0 +1,61 @@
+from django.db import models
+from django.core.exceptions import ValidationError
+from django.db.models import Q
+
+from django.contrib.auth.models import User
+from course.models import TeacherCourses, StudentCourses
+
+
+class PostManager(models.Manager):
+    @staticmethod
+    def get_posts_for_user(user: User):
+        relevant_users = StudentCourses.objects.filter(
+            Q(Q(student_id=user) & Q(status="Confirmed") | Q(teacher_course_id__teacher_id=user))
+        )
+        relevant_courses = TeacherCourses.objects.filter(studentcourses__in=relevant_users).distinct()
+        relevant_posts = Post.objects.filter(course_id__in=relevant_courses).order_by("date")
+
+        # create the post hierarchy in a dictionary that contains lists
+        post_hierarchy = {}
+        for post in relevant_posts:
+            if post.parent_post_id:
+                post_hierarchy[post.parent_post_id].append(post)
+            else:
+                post_hierarchy[post] = []
```

Comment:
Maybe you can get this with a filter? something like you can do in SQL `WHERE parent_post_id==post`

Reply:
I didn’t want to do additional queries, Would this option be preferable?
If so, I think I can access related replies by using something like `post.post__set`

Reply:
Yes, I meant to use something built-in instead of doing this for loop

Reply:
I think you can do something like `relevant_posts.filter(parent_post_id= post.parent_post_id)` to get all the posts with this parent. I wonder if this functionality should be part of this method or somewhere else?

Reply:
`relevant_posts` queryset contains all posts that are tied to user (posts that belong to a course the user is signed to/ owner of) regardless of whether post is a root post, or a reply post.

What I thought there is left to do is to organize the posts in a dictionary of lists where the key is a root post and the value is a list of replies (sorted by least recent to most recent reply). I thought this would be the most convenient data structure as it mimics how it should be displayed in the feed view.

maybe I still don't understand correctly what you mean,
but if I fetch for each post its replies using a filter it will lead to an additional query for each post (unless it gets optimized in a way I'm unaware of)

Reply:
Got it, thanks